arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Resource.kt

```diff
@@ -464,12 +464,48 @@ public sealed class Resource<out A> {
   public fun <B, C> parZip(
     ctx: CoroutineContext = Dispatchers.Default,
     fb: Resource<B>,
-    f: suspend (A, B) -> C,
+    f: suspend (A, B) -> C
   ): Resource<C> =
     arrow.fx.coroutines.continuations.resource {
       parZip(ctx, { this@Resource.bind() }, { fb.bind() }) { a, b -> f(a, b) }
     }
 
+  /**
+   * Decomposes a [Resource]<A> into a Pair<suspend () -> A, suspend (A, ExitCase) -> Unit>, containing
+   * the allocation and release functions which make up the [Resource].
+   *
+   * This can be used to integrate Resources with code which cannot be run within the [use] function.
+   */
```

Comment:
I think we need to be more elaborate in this piece of documentation. It's a very delicate API, and should also carefully explain its correct usage.

```suggestion
  /**
   * Deconstruct [Resource] into an `acquire` and `release` handlers.
   * The `release` action **must** always be called with resource [A] returned from `acquire`, if the `release` step is never called, then the resource [A] will leak. The `acquire` and `release` steps are already made `NonCancellable` to guarantee correct invocation like `Resource` or `bracketCase`.
   *
   * ```kotlin
   * val resource = Resource({ println("Acquire") }) { _, exitCase ->
   *  println("Release $exitCase")
   * }
   * 
   * suspend fun main(): Unit {
   *   val (acquire, release) = resource.allocated()
   *   val a = acquire()
   *   try {
   *     /** Do something with A */
   *     release(a, ExitCase.Completed)
   *   } catch(e: Throwable) {
   *      val e2 = runCatching { release(a, ExitCase(e)) }.exceptionOrNull()
   *      throw Platform.composeErrors(e, e2)
   *   }
   *}
   * ```
   * 
   * This is a **delicate** API. It is easy to accidentally create resource or memory leaks `allocated` is used.
   * A `Resource` allocated by `allocated` is not subject to the guarantees that [Resource] makes,
   * instead the caller is responsible of correctly invoking the `release` handler at the appropriate time.
   * This API is useful for building inter-op APIs between [Resource] and non-suspending code, such as Java libraries.
   */
```

While write this doc, we should probably expose an smart-constructor for `val exit = if(e is CancellationException) ExitCase.Cancelled(e) else ExitCase.Failure(e)`. Do you mind adding it in this PR? We're also using it several times inside this PR. Something like the following function heavily cleans up this user-example above, and it would also clean-up several instances in our internal Resource code I think.

```kotlin
/** Construct an ExitCase from a [Throwable] */
public fun ExitCase(error: Throwable): ExitCase =
  if (error is CancellationException) ExitCase.Cancelled(error) else ExitCase.Failure(error)
```